lib/core/proxy.c

```diff
@@ -398,11 +442,35 @@ static inline void on_websocket_upgrade(struct rp_generator_t *self, h2o_timeout
     h2o_req_t *req = self->src_req;
     h2o_socket_t *sock = h2o_http1client_steal_socket(self->client);
     h2o_buffer_consume(&sock->input, rlen); // trash data after stealing sock.
-    struct rp_ws_upgrade_info_t *info = h2o_mem_alloc(sizeof(*info));
-    info->upstream_sock = sock;
-    info->timeout = timeout;
-    info->ctx = req->conn->ctx;
-    h2o_http1_upgrade(req, NULL, 0, on_websocket_upgrade_complete, info);
+
+    if (req->version < 0x200) {
+        h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_UPGRADE, NULL, H2O_STRLIT("websocket"));
+        struct rp_ws_upgrade_info_t *info = h2o_mem_alloc(sizeof(*info));
+        info->upstream_sock = sock;
+        info->timeout = timeout;
+        info->ctx = req->conn->ctx;
+        h2o_http1_upgrade(req, NULL, 0, on_websocket_upgrade_complete, info);
```

Comment:
What happens if a mruby handler creates a request that looks like a websocket request? I would assume that something very strange will happen.

See my response below for the generic abstraction that me might want to use.

Reply:
> I would assume that something very strange will happen.

What happens do you think?

> See my response below for the generic abstraction that me might want to use.

Which response?